src/api/dashboard.js

```diff
@@ -56,14 +63,29 @@ Dashboard.prototype = {
     this._dashboard.vis.mapvis.map.setBounds([state.map.ne, state.map.sw]);
   },
 
-  onStateChanged: function (callback, shareURLs) {
-    this._dashboard.vis.once('dataviewsFetched', function () {
-      this._dashboard.widgets._widgetsCollection.initialState();
-      this._dashboard.widgets._widgetsCollection.each(function (m) {
-        m.applyInitialState();
-      }, this);
-      shareURLs === true && this._bindChange(callback);
-    }, this);
+  _widgetsAlreadyInitialized: function () {
+    var widgetsCollection = this._dashboard.widgets._widgetsCollection;
+    if (widgetsCollection.size() > 0) {
+      return this._dashboard.widgets._widgetsCollection.hasInitialState();
+    }
+    return true;
+  },
+
+  onDataviewsFetched: function (callback) {
+    var areDataviewsFetched = this._widgetsAlreadyInitialized();
+    if (areDataviewsFetched) {
+      callback && callback();
+    } else {
+      this._dashboard.vis.once('dataviewsFetched', function () {
```

Comment:
Well, not that agree. If we make this change we could end up in some confusions. In DI we only need the first event triggered, but in CARTO.js we are sending the same event when all dataviews are fetched. So that means the "binding" should be decided in the DI field (bind, once, stop, off, on, whatever we want)... I'd leave it for now as it is.